pkg/registry/apps/statefulset/strategy_test.go

```diff
@@ -22,76 +22,100 @@ import (
+				PersistentVolumeClaimPolicy: &apps.StatefulSetPersistentVolumeClaimPolicy{
+					OnSetDeletion: apps.DeletePersistentVolumeClaimDeletePolicyType,
+					OnScaleDown:   apps.RetainPersistentVolumeClaimDeletePolicyType,
+				},
 			},
-		},
-	}
-	ps := &apps.StatefulSet{
-		ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
-		Spec: apps.StatefulSetSpec{
-			PodManagementPolicy: apps.OrderedReadyPodManagement,
-			Selector:            &metav1.LabelSelector{MatchLabels: validSelector},
-			Template:            validPodTemplate.Template,
-			UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
-		},
-		Status: apps.StatefulSetStatus{Replicas: 3},
-	}
+			Status: apps.StatefulSetStatus{Replicas: 3},
+		}
 
-	Strategy.PrepareForCreate(ctx, ps)
-	if ps.Status.Replicas != 0 {
-		t.Error("StatefulSet should not allow setting status.replicas on create")
-	}
-	errs := Strategy.Validate(ctx, ps)
-	if len(errs) != 0 {
-		t.Errorf("unexpected error validating %v", errs)
-	}
+		Strategy.PrepareForCreate(ctx, ps)
+		if ps.Status.Replicas != 0 {
+			t.Error("StatefulSet should not allow setting status.replicas on create")
+		}
+		errs := Strategy.Validate(ctx, ps)
+		if len(errs) != 0 {
+			t.Errorf("unexpected error validating %v", errs)
+		}
 
-	// Just Spec.Replicas is allowed to change
-	validPs := &apps.StatefulSet{
-		ObjectMeta: metav1.ObjectMeta{Name: ps.Name, Namespace: ps.Namespace, ResourceVersion: "1", Generation: 1},
-		Spec: apps.StatefulSetSpec{
-			PodManagementPolicy: apps.OrderedReadyPodManagement,
-			Selector:            ps.Spec.Selector,
-			Template:            validPodTemplate.Template,
-			UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
-		},
-		Status: apps.StatefulSetStatus{Replicas: 4},
-	}
-	Strategy.PrepareForUpdate(ctx, validPs, ps)
-	errs = Strategy.ValidateUpdate(ctx, validPs, ps)
-	if len(errs) != 0 {
-		t.Errorf("updating spec.Replicas is allowed on a statefulset: %v", errs)
-	}
+		// Just Spec.Replicas is allowed to change
+		validPs := &apps.StatefulSet{
+			ObjectMeta: metav1.ObjectMeta{Name: ps.Name, Namespace: ps.Namespace, ResourceVersion: "1", Generation: 1},
+			Spec: apps.StatefulSetSpec{
+				PodManagementPolicy: apps.OrderedReadyPodManagement,
+				Selector:            ps.Spec.Selector,
+				Template:            validPodTemplate.Template,
+				UpdateStrategy:      apps.StatefulSetUpdateStrategy{Type: apps.RollingUpdateStatefulSetStrategyType},
+			},
+			Status: apps.StatefulSetStatus{Replicas: 4},
+		}
+		// The PersistentVolumeClaimPolicy should only be present if the feature gate is enabled.
+		if utilfeature.DefaultFeatureGate.Enabled(features.StatefulSetAutoDeletePVC) {
+			validPs.Spec.PersistentVolumeClaimPolicy = &apps.StatefulSetPersistentVolumeClaimPolicy{
+				OnSetDeletion: apps.DeletePersistentVolumeClaimDeletePolicyType,
```

Comment:
Are these values mutable? If so, check that they can be changed.

Reply:
done.